src/Application/Kernel.php

```diff
@@ -161,4 +149,57 @@ private function initDiscovery(Container $container): void
             $discovery->storeCache();
         }
     }
+
+    private function discoverInstalledPackageLocations(): void
+    {
+        $composerPath = path($this->root, 'vendor/composer');
+        $installed = path($composerPath, 'installed.json');
+
+        if (is_file($installed)) {
+            $installed = $this->loadJsonFile($installed);
+
+            foreach ($installed['packages'] ?? [] as $package) {
+                $packagePath = realpath(path($composerPath, $package['install-path']));
+
+                if (isset($package['require']['tempest/framework'], $package['autoload']['psr-4'])
```

Comment:
I would refactor this for clarity into two boolean variables:

```php
$requiresTempest = isset($package['require']['tempest/framework']);
$hasRootNamespaces = isset($package['autoload']['psr-4']) && is_array($package['autoload']['psr-4']);

if ($requiresTempest && $hasRootNamespaces) {
```

Reply:
I don't think we need the check for whether it's an array.

Reply:
True, composer already validates that

Reply:
Agreed on removing `is_array()` ... i'm too accustomed to code with unreliable data structures in it. i should _hope_ composer is more thoughtful than that. 😉 

As for separating the `isset()` ... i agree the variables describe the condition, and i'll make the change.

For what it's worth, i try to balance concise code with descriptive code. Separate variables, each calling `isset()`, felt like variable pollution. Whereas one call that checks both felt sufficient. Personal preference, i don't feel strongly enough about it to protest. 😃